Tests/AsyncQueueTests/ActorQueueTests.swift

```diff
@@ -0,0 +1,184 @@
+// MIT License
+//
+// Copyright (c) 2022 Dan Federman
+//
+// Permission is hereby granted, free of charge, to any person obtaining a copy
+// of this software and associated documentation files (the "Software"), to deal
+// in the Software without restriction, including without limitation the rights
+// to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+// copies of the Software, and to permit persons to whom the Software is
+// furnished to do so, subject to the following conditions:
+//
+// The above copyright notice and this permission notice shall be included in all
+// copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
+// SOFTWARE.
+
+import XCTest
+
+@testable import AsyncQueue
+
+final class ActorQueueTests: XCTestCase {
+
+    // MARK: XCTestCase
+
+    override func setUp() async throws {
+        try await super.setUp()
+
+        systemUnderTest = ActorQueue()
+    }
+
+    // MARK: Behavior Tests
+
+    func test_async_sendsEventsInOrder() async {
+        let counter = Counter()
+        for iteration in 1...1_000 {
+            systemUnderTest.async {
+                await counter.incrementAndExpectCount(equals: iteration)
+            }
+        }
+        await systemUnderTest.await { /* Drain the queue */ }
+    }
+
+    func test_async_sendsSynchronousEventsInOrder() async {
+        final class SynchronousCounter: @unchecked Sendable {
+            func incrementAndExpectCount(equals expectedCount: Int) {
+                increment()
+                XCTAssertEqual(expectedCount, count)
+            }
+
+            func increment() {
+                count += 1
+            }
+
+            var count: Int {
+                set {
+                    dispatchQueue.async { self.unsafeCount = newValue }
+                }
+                get {
+                    dispatchQueue.sync { unsafeCount }
+                }
+            }
+            var unsafeCount = 0
+            let dispatchQueue = DispatchQueue(label: #function)
+        }
+        let counter = SynchronousCounter()
+        for iteration in 1...1_000 {
+            systemUnderTest.async {
+                counter.incrementAndExpectCount(equals: iteration)
+            }
+        }
+        await systemUnderTest.await { /* Drain the queue */ }
+    }
+
+    func test_async_startsExecutionOfNextTaskAfterSuspension() async {
```

Comment:
I could be missing something but would this test also pass with a `FIFOQueue`? I'm trying to think if there's a more targeted way to verify/show that the next task starts to execute after the first task suspends. It may be that we don't need 1000 tasks for this type of test, and just want a couple?

Reply:
This test deadlocks when the `systemUnderTest` is a `FIFOQueue` instance.

The reason it deadlocks is:
1. A `systemUnderTest.async` task `await`s the semaphore
2. A future `systemUnderTest.async` task signals the semaphore, but that future task never executes because the first task is stuck waiting for the semaphore to unlock

Reply:
Got it. Maybe we can add some comments connecting the dots in the test code for how we are proving that a task starts before the previous task finishes?

Reply:
Good call! Will do on the next pass

Reply:
583ca33